scripts/joint_goal_publisher.py

```diff
@@ -0,0 +1,338 @@
+                        minval = float(limit.getAttribute('lower'))
+                        maxval = float(limit.getAttribute('upper'))
+                    except:
+                        rospy.logwarn("%s is not fixed, nor continuous, but limits are not specified!" % name)
+                        continue
+
+                safety_tags = child.getElementsByTagName('safety_controller')
+                if self.use_small and len(safety_tags) == 1:
+                    tag = safety_tags[0]
+                    if tag.hasAttribute('soft_lower_limit'):
+                        minval = max(minval, float(tag.getAttribute('soft_lower_limit')))
+                    if tag.hasAttribute('soft_upper_limit'):
+                        maxval = min(maxval, float(tag.getAttribute('soft_upper_limit')))
+
+                mimic_tags = child.getElementsByTagName('mimic')
+                if self.use_mimic and len(mimic_tags) == 1:
+                    tag = mimic_tags[0]
+                    entry = {'parent': tag.getAttribute('joint')}
+                    if tag.hasAttribute('multiplier'):
+                        entry['factor'] = float(tag.getAttribute('multiplier'))
+                    if tag.hasAttribute('offset'):
+                        entry['offset'] = float(tag.getAttribute('offset'))
+
+                    self.dependent_joints[name] = entry
+                    continue
+
+                if name in self.dependent_joints:
+                    continue
+
+                if self.zeros and name in self.zeros:
+                    zeroval = self.zeros[name]
+                elif minval > 0 or maxval < 0:
+                    zeroval = (maxval + minval)/2
+                else:
+                    zeroval = 0
+
+                joint = {'min': minval, 'max': maxval, 'zero': zeroval}
+                #if self.pub_def_positions:
+                  #  joint['position'] = zeroval
+                #if self.pub_def_vels:
+                 #   joint['velocity'] = 0.0
+                #if self.pub_def_efforts:
+                 #   joint['effort'] = 0.0
+
+                if jtype == 'continuous':
+                    joint['continuous'] = True
+                self.free_joints[name] = joint
+
+    def send_goal(self, goal):
+        """
+        sends a joint goal to giskard
+        :param goal:
+        :type goal: dict
+        :return:
+        """
+        self.giskard_wrapper.set_joint_goal(goal)
+        self.giskard_wrapper.plan_and_execute()
+
+
+    def __init__(self):
+        description = get_param('robot_description')
+
+        self. giskard_wrapper = GiskardWrapper()
+
+        self.free_joints = {}
+        self.joint_list = [] # for maintaining the original order of the joints
+        self.dependent_joints = get_param("dependent_joints", {})
+        self.use_mimic = get_param('use_mimic_tags', True)
+        self.use_small = get_param('use_smallest_joint_limits', True)
```

Comment:
setting defaults for ros params is usually a bad idea, because then a user doesn't know which params exist for a node. Better to have no defaults in the code and instead setting defaults in a launch file.

Reply:
I left the default values in the code because otherwise the script would crash if you run it without the launch file. Now you can either use the launch file and set the default values that way or you can run it without the launch file and it will use the default values in the code.